### ISSUE TITLE
P4-14 converter and parallel actions

### ISSUE THREAD

**Issue by jnfoster** [p4lang/p4c#246]
The P4-14 spec states,
> P4 assumes parallel semantics for the application of all the primitive actions executing as a result of a match in a given table.
and gives the following example,
```
modify_field(hdr.fieldA, 1);
modify_field(hdr.fieldB, hdr.fieldA);
```
further clarifying that: 
> With parallel semantics, both actions are started at the same time, so the evaluation of hdr.fieldA in the second instruction resolves to 0 (since it has not yet changed), and so hdr.fieldB receives the
value 0.

However, if I compile this program 
```
header_type hdr_t {
    fields {
        fieldA : 32;
        fieldB : 32;
    }
}

header hdr_t hdr;

parser start {
    return parse_h;
}

parser parse_h {
    extract(hdr);
    return ingress;
}

action a() {
    modify_field(hdr.fieldA, 1);
    modify_field(hdr.fieldB, hdr.fieldA);
}

table t {
    reads { }
    actions { a; }
    size : 512;
}

control ingress{
    apply(t);
}
```
using the P4-14 to P4-16 converter, I get a result that does not execute correctly under sequential semantics:
```
#include <core.p4>
#include <v1model.p4>

header hdr_t {
    bit<32> fieldA;
    bit<32> fieldB;
}

struct metadata {
}

struct headers {
    @name("hdr") 
    hdr_t hdr;
}

parser ParserImpl(packet_in packet, out headers hdr, inout metadata meta, inout standard_metadata_t standard_metadata) {
    @name("parse_h") state parse_h {
        packet.extract(hdr.hdr);
        transition accept;
    }
    @name("start") state start {
        transition parse_h;
    }
}

control ingress(inout headers hdr, inout metadata meta, inout standard_metadata_t standard_metadata) {
    @name("a") action a() {
        hdr.hdr.fieldA = 32w1;
        hdr.hdr.fieldB = hdr.hdr.fieldA;
    }
    @name("t") table t() {
        actions = {
            a;
            NoAction;
        }
        key = {
        }
        size = 512;
        default_action = NoAction();
    }
    apply {
        t.apply();
    }
}

control egress(inout headers hdr, inout metadata meta, inout standard_metadata_t standard_metadata) {
    apply {
    }
}

control DeparserImpl(packet_out packet, in headers hdr) {
    apply {
        packet.emit(hdr.hdr);
    }
}

control verifyChecksum(in headers hdr, inout metadata meta) {
    apply {
    }
}

control computeChecksum(inout headers hdr, inout metadata meta) {
    apply {
    }
}

V1Switch(ParserImpl(), verifyChecksum(), ingress(), egress(), computeChecksum(), DeparserImpl()) main;
```

**Comment by ChrisDodd** [comment]
That's because we are following the v1.1 P4_14 spec which was revised to specify sequential semantics, in accordance with the existing implementations at the time.

**Comment by mbudiu-vmw** [comment]
Unfortunately the P4 1.1 spec has been retired.
But this is a bug which I won't fix - the spec is just plain wrong (ambiguous).

**Comment by jnfoster** [comment]
I don't know the past history. And yes, the spec is vague, but it gives
this very example and yet compiling with p4c gives a different answer.

-N

**Comment by cc10512** [comment]
We should consider carefully implementing a consistent behavior across different versions of the compiler. We can not afford to have the _same program_ compiled with the same compiler producing different results, even when the compilation goes through a translation from one semantics to another. At the least, the P4_14-> P4_16 converter should raise an error since the behavior changes. Thus, the P4_14 compilation can continue to use parallel semantics, and if the code is written in P4_16, it will follow sequential semantics. However, the translation should require programmer intervention.

**Comment by mbudiu-vmw** [comment]
The main problem with the "parallel semantics" - assuming it was defined carefully - is language usability. I am willing to bet a tidy sum that people will hate using such a language. Moreover, there are lots of useful programs which simply cannot be written - especially if you only consider using the original P4-14 spec which had no support for arbitrary expressions. Third - programmers don't read specs, so this behavior will come as a surprise to almost anyone. The syntax does not indicate the parallelism in any way. Fourth, some of the programs we already have in P4-14 require sequential semantics. So it is not obvious at all how to differentiate between the two "dialects" of P4-14 now in existence. 

The effort to support parallel semantics is just an an academic exercise - there will be no new programs written using parallel semantics. We could support this by adding a special compiler flag e.g., --parallel, to have users indicate parallel P4-14 semantics. I currently have a big enough backlog so that I will give very low priority to this issue. If someone is interested in working on it, the right place to do it is probably the SideEffectOrdering pass in the front-end. It's more interesting and useful to work on the opposite transformation, to parallelize code in the back-ends which require it.

**Comment by jnfoster** [comment]
I agree this is low priority. I just want to get the spec and open-source
tools in order.

Maybe simplest would be to revise the P4-14 1.03 spec to eliminate the
discussion of parallel semantics.

-N

**Comment by chkim4142** [comment]
+1 for changing the P4_14 v1.0.3 spec.

**Comment by gbrebner** [comment]
+1 for revising the v1.0.3 spec.

**Comment by mbudiu-vmw** [comment]
This issue should have been solved by the P4 v1.0.3 spec. Our compiler follows that version of the spec in this respect.
